sql/parser/builtins.go

```diff
@@ -72,6 +74,67 @@ var builtins = map[string]builtin{
 	"upper": stringBuiltin(func(s string) (Datum, error) {
 		return DString(strings.ToUpper(s)), nil
 	}),
+
+	// TODO(XisiHuang): support the substring(str FROM x FOR y) syntax
+	"substring": {
+		nArgs: -1,
+		fn: func(args DTuple) (Datum, error) {
+			argsNum := len(args)
+			if argsNum != 2 && argsNum != 3 {
+				return DNull, fmt.Errorf("incorrect number of arguments: %d vs %s",
+					argsNum, "2 or 3")
+			}
+
+			dstr, ok := args[0].(DString)
+			if !ok {
+				return DNull, argTypeError(args[0], "string")
+			}
+			str := string(dstr)
+
+			start, ok := args[1].(DInt)
+			if !ok {
+				return DNull, argTypeError(args[1], "int")
+			}
+
+			s := int(start)
+			if s < 0 || s > len(str)-1 {
+				return DNull, invalidArgError(fmt.Sprintf("%d is out of bounds", s))
+			}
+
+			var c int
+			if argsNum == 2 {
+				c = len(str) - s
+			} else {
+				count, ok := args[2].(DInt)
+				if !ok {
+					return DNull, argTypeError(args[2], "int")
+				}
+				c = int(count)
+				if c < 0 || s+c > len(str) {
+					return DNull, invalidArgError(fmt.Sprintf("%d is out of bounds", c))
```

Comment:
Let's tweak this error message to: `fmt.Sprintf("%d+%d > len(%s) [%d]", s, c, len(str), str)`